File: dotnet/Falcon.PluginAPI/Items/PlayerInventory.cs
using Falcon.Interop;

namespace Falcon
{
    public sealed unsafe class PlayerInventory
    {
        private readonly FalconPlayer* _player;

        internal PlayerInventory(FalconPlayer* player)
        {
            _player = player;
        }

        public uint Size
        {
            get
            {
                return NativeApi.Table->playerInventorySize(_player);
            }
        }

        public uint SelectedSlot
        {
            get
            {
                return NativeApi.Table->playerSelectedSlot(_player);
            }
            set
            {
                NativeApi.Table->playerSetSelectedSlot(_player, value);
            }
        }

        public Item? HeldItem
        {
            get
            {
                return GetItem(SelectedSlot);
            }
        }

        public Item? Offhand
        {
            get
            {
                return Item.Adopt(NativeApi.Table->playerOffhandItem(_player));
            }
        }

        public Item? GetItem(uint slot)
        {
            return Item.Adopt(NativeApi.Table->playerInventoryItem(_player, slot));
        }

        public void SetItem(uint slot, Item item)
        {
            NativeApi.Table->playerSetInventoryItem(_player, slot, item.Raw);
        }

        public bool Give(Item item)
        {
            return NativeApi.Table->playerGiveItem(_player, item.Raw) != 0;
        }

        public Item? GetArmor(ArmorSlot slot)
        {
            return Item.Adopt(NativeApi.Table->playerArmorItem(_player, (uint)slot));
        }

        public void SetArmor(ArmorSlot slot, Item item)
        {
            NativeApi.Table->playerSetArmorItem(_player, (uint)slot, item.Raw);
        }

        public void SetOffhand(Item item)
        {
            NativeApi.Table->playerSetOffhandItem(_player, item.Raw);
        }

        public void Clear()
        {
            NativeApi.Table->playerClearInventory(_player);
        }
    }
}

File: java/src/main/java/falcon/api/PlayerInventory.java
package falcon.api;

import falcon.api.internal.Interop;
import falcon.api.internal.NativeApi;

import java.lang.foreign.MemorySegment;
import java.util.Optional;

public final class PlayerInventory {
    private final MemorySegment mPlayer;

    public PlayerInventory(MemorySegment player) {
        mPlayer = player;
    }

    public int size() {
        return api().playerInventorySize(mPlayer);
    }

    public Optional<Item> item(int slot) {
        return Item.adopt(api().playerInventoryItem(mPlayer, slot));
    }

    public void setItem(int slot, Item item) {
        api().playerSetInventoryItem(mPlayer, slot, item.handle());
    }

    public boolean give(Item item) {
        return api().playerGiveItem(mPlayer, item.handle()) != 0;
    }

    public int selectedSlot() {
        return api().playerSelectedSlot(mPlayer);
    }

    public void setSelectedSlot(int slot) {
        api().playerSetSelectedSlot(mPlayer, slot);
    }

    public Optional<Item> heldItem() {
        return item(selectedSlot());
    }

    public Optional<Item> armor(ArmorSlot slot) {
        return Item.adopt(api().playerArmorItem(mPlayer, slot.value()));
    }

    public void setArmor(ArmorSlot slot, Item item) {
        api().playerSetArmorItem(mPlayer, slot.value(), item.handle());
    }

    public Optional<Item> offhand() {
        return Item.adopt(api().playerOffhandItem(mPlayer));
    }

    public void setOffhand(Item item) {
        api().playerSetOffhandItem(mPlayer, item.handle());
    }

    public void clear() {
        api().playerClearInventory(mPlayer);
    }

    private static NativeApi api() {
        return Interop.api();
    }
}
